File: s.js
class ChessGame {
    constructor() {
        this.initialize_board();
        this.current_player = 'White';
        this.flags = [['e1', 'White', 0], ['e8', 'Black', 0]];
        this.base = [['d1', 'White', 0], ['d8', 'Black', 0]];
        this.piece = ' ';
        this.piece_id = ' ';
        this.select = null;
        this.print_board();
        const cells = document.querySelectorAll('#chessboard td');
        cells.forEach(cell => {
            cell.addEventListener('click', () => {

                this.validate_move(cell)
            });
        });
    }

    validate_move(cell) {
        const end = this.convert_position(cell.id);
        if (this.select == cell) {
            cell.classList.remove("selected");
            this.select = null;
            return;
        }
        else if (this.select == null) {
            this.select = cell
            cell.classList.add("selected");
        }
        else {
            var pos = this.convert_position(this.select.id)
            var pos_to = this.convert_position(cell.id)
            if (this.is_valid_move(this.board[pos[0]][pos[1]][1], pos[0], pos[1], pos_to[0], pos_to[1])) {
                this.move_piece(this.board[pos[0]][pos[1]][0], pos[0], pos[1], pos_to[0], pos_to[1])
            }
            else {
                document.getElementById(this.select.id).classList.remove("selected");
                this.select = null
            }
        }
    }

    initialize_board() {
        this.board = Array.from({ length: 8 }, () => Array.from({ length: 8 }, () => null));

        this.board[0] = [['♖', 'R', 'Black', 0], ['♘', 'N', 'Black', 0], ['♗', 'B', 'Black', 0], ['🎯', 'C', 'Black', 0], ['🏴', 'K', 'Black', 0], ['♗', 'B', 'Black', 0], ['♘', 'N', 'Black', 0], ['♖', 'R', 'Black', 0]];
        this.board[1] = [['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0]];

        this.board[5] = [[' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0]];
        this.board[4] = [[' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0]];
        this.board[3] = [[' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0]];
        this.board[2] = [[' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0], [' ', '.', '.', 0]];


        this.board[7] = [['♜', 'r', 'White', 0], ['♞', 'n', 'White', 0], ['♝', 'b', 'White', 0], ['🎯', 'c', 'White', 0], ['🏳️', 'k', 'White', 0], ['♝', 'b', 'White', 0], ['♞', 'n', 'White', 0], ['♜', 'r', 'White', 0]];
        this.board[6] = [['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0]];

    }

    print_board() {
        let table = document.createElement('table');
        table.setAttribute('border', '1');

        for (let row = 7; row >= 0; row--) {
            let tr = document.createElement('tr');
            for (let col = 7; col >= 0; col--) {
                let td = document.createElement('td');
                if (this.board[row][col] === undefined) {
                    this.board[row][col] = [' ', '.', '.', 0]
                }
                let pieceSymbol = this.board[row][col] ? this.board[row][col][0] : ' ';
                td.textContent = pieceSymbol;
                if (this.board[row][col][3] == 2) {
                    td.style.color = "blue!important"
                }
                if (this.board[row][col][3] == 1) {
                    td.style.color = "red!important"
                }
                td.id = `${String.fromCharCode(row + 'a'.charCodeAt())}` + `${col}`;
                td.classList.add('start');
                tr.appendChild(td);
            }
            table.appendChild(tr);
        }

        document.getElementById("chessboard").appendChild(table);
    }

    capture_piece(row, col, moving_piece) {
        if (this.board[row][col] !== null) {
            this.board[row][col] = null;
        }
    }

    restore_spot(end_row, end_col) {
        return String.fromCharCode(`${7 - end_row + 'a'.charCodeAt()}`) + `${7 - end_col}`;
    }

    move_piece(piece, start_row, start_col, end_row, end_col) {
        console.log(end_row);
        console.log(String.fromCharCode(`${end_row + 'a'.charCodeAt()}`) + `${(7 + end_col) % 8}`);
        document.getElementById(String.fromCharCode(`${7 - end_row + 'a'.charCodeAt()}`) + `${7 - end_col}`).textContent = piece;
        var base = document.getElementById(String.fromCharCode(`${7 - end_row + 'a'.charCodeAt()}`) + `${7 - end_col}`)
        if (this.board[end_row][end_col][1].toLowerCase() == 'c') {
            this.board[start_row][start_col][3] = (this.board[start_row][start_col][2] == this.board[end_row][end_col][2]) ? 1 : 2;
        }
        // [3] is 1 when own flag when being returned [3] is 2 when bringing back other teams
        if (base.id == 'a3' && this.board[start_row][start_col][3] == 2) {
            window.alert("White Wins!!")
            document.getElementById("chessboard").innerHTML = ""
            this.initialize_board();
            this.print_board()
        }
        else if (base.id == 'h3' && this.board[start_row][start_col][3] == 2) {
            window.alert("Black Wins!!")
            document.getElementById("chessboard").innerHTML = ""
            this.initialize_board();
            this.print_board()
        }
        else if (base.id == 'h3' && this.board[start_row][start_col][3] == 1) {
            window.alert("Black Home!!")
            document.getElementById(String.fromCharCode(`${7 - end_row + 'a'.charCodeAt()}`) + `${7 - end_col}`).textContent = ' ';
            document.getElementById('h3').textContent = '🎯'
            document.getElementById('h4').textContent = '🏳️'
        }
        else if (base.id == 'a3' && this.board[start_row][start_col][3] == 1) {
            window.alert("White Home!!")
            document.getElementById(String.fromCharCode(`${7 - end_row + 'a'.charCodeAt()}`) + `${7 - end_col}`).textContent = ' ';
            document.getElementById('a3').textContent = '🎯'
            document.getElementById('a4').textContent = '🏴'
        }
        else {
            this.board[end_row][end_col] = this.board[start_row][start_col];
            document.getElementById(this.select.id).textContent = ' ';
        }
        document.getElementById(this.select.id).classList.remove("selected");
        this.select = null
    }

    convert_position(position) {
        const col = Math.abs(7 - parseInt(position[1]));
        const row = 7 - Math.abs(position.charCodeAt(0) - 'a'.charCodeAt(0));
        return [row, col];
    }

    is_valid_move(piece, start_row, start_col, end_row, end_col) {
        if (piece.toLowerCase() == 'c' || piece.toLowerCase() == 'k')
            return false;
        // Check for obstruction in the path
        this.piece = piece
        if (this.piece === 'R' || this.piece === 'r') {
            if (start_row === end_row) { // Horizontal move
                const delta_col = end_col > start_col ? 1 : -1;
                for (let col = start_col + delta_col; col !== end_col; col += delta_col) {
                    if (this.board[start_row][col] === null) {
                        this.board[start_row][col] = [' ', '.', '.', 0];
                    }
                    if (this.board[start_row][col][1] !== '.') {
                        return false; // Obstruction found
                    }
                }
            } else if (start_col === end_col) { // Vertical move
                const delta_row = end_row > start_row ? 1 : -1;
                for (let row = start_row + delta_row; row !== end_row; row += delta_row) {
                    if (this.board[row][start_col] === null) {
                        this.board[row][start_col] = [' ', '.', '.', 0];
                    } 
                    if (this.board[row][start_col][1] !== '.') {
                        return false; // Obstruction found
                    }
                }
            }
            return true;
        }

        if (this.piece === 'P' || this.piece === 'p') { // Pawn or King
            // Check for edge cases
            if ((start_row === 0 && end_row === 7 && start_col === end_col) ||
                (start_row === 7 && end_row === 0 && start_col === end_col) ||
                (start_col === 0 && end_col === 7 && start_row === end_row) ||
                (start_col === 7 && end_col === 0 && start_row === end_row)) {
                return false;
            }
        }

        // Other piece-specific move validations
        if (this.piece === 'N' || this.piece === 'n') { // Knight
            const delta_row = Math.abs(end_row - start_row);
            const delta_col = Math.abs(end_col - start_col);
            console.log("*" + delta_col + " " + delta_row)
            if (delta_row == 2 && delta_col == 1)
                return true;
            else if (delta_row == 1 && delta_col == 2)
                return true
            else if (delta_row == 1 && delta_col >= 6)
                return true
            else if (delta_row >= 6 && delta_col == 1)
                return true
            else
                return false
        }
        if (this.piece === 'P' || this.piece === 'p') { // Pawn
            // Pawn specific move validations
            if (start_row === 1 || start_row == 6 && start_col === end_col && Math.abs(end_row - start_row) === 2) {
                return true;
            }
            if (start_col === end_col && Math.abs(end_row - start_row) === 1) {
                return true;
            }
            if (this.board[end_row][end_col] !== null && this.board[start_row][start_col][2] !== this.board[end_row][end_col][2] && Math.abs(start_col - end_col) === 1 && Math.abs(end_row - start_row) === 1) {
                return true;
            }
            return false;
        } else if (this.piece === 'R' || this.piece === 'r') { // Rook
            // Rook moves horizontally or vertically
            return start_row == end_row || start_col == end_col;
        } else if (this.piece === 'B' || this.piece === 'b') { // Bishop
            // Bishop moves diagonally
            return Math.abs(end_row - start_row) === Math.abs(end_col - start_col);
        } else {
            return false; // Default: invalid move
        }
    }

    is_check(piece, start_row, start_col, flags_row, flags_col) {
        if (!(0 <= start_row < 8 && 0 <= start_col < 8 && 0 <= flags_row < 8 && 0 <= flags_col < 8)) {
            return false;
        }

        if (this.board[start_row][start_col] === null) {
            return false;
        }

        if (this.piece === 'R' || this.piece === 'r') { // Rook or Queen
            if (start_row === flags_row) { // Horizontal move
                for (let i = 0; i < 8; i++) {
                    if (this.board[start_row][i] === null) {
                        this.board[start_row][i] = ['.', '.', '.'];
                    }
                    if (this.board[start_row][i][2] !== this.board[start_row][i][2] && this.board[start_row][i][1].toLowerCase() !== 'k') {
                        return false;
                    }
                    if (this.board[start_row][i][2] !== this.board[start_row][i][2] && this.board[start_row][i][1].toLowerCase() === 'k') {
                        return true;
                    }
                }
            } else if (start_col === flags_col) { // Vertical move
                for (let i = 0; i < 8; i++) {
                    if (this.board[i][start_col] === null) {
                        this.board[i][start_col] = ['.', '.', '.'];
                    }
                    if (this.board[i][start_col][2] !== this.board[start_row][start_col][2] && this.board[i][start_col][1].toLowerCase() !== 'k') {
                        return false;
                    }
                    if (this.board[i][start_col][2] !== this.board[start_row][start_col][2] && this.board[i][start_col][1].toLowerCase() === 'k') {
                        return true;
                    }
                }
            }
        }

        if (this.piece === 'B' || this.piece === 'b') { // Bishop or Queen
            for (let i = 0; i < 8; i++) {
                for (let j = 0; j < 8; j++) {
                    if (j === i) {
                        if (this.board[i][j] === null) {
                            this.board[i][j] = ['.', '.', '.'];
                        }
                        if (this.board[i][j][2] === this.board[start_row][start_col][2]) {
                            break;
                        }
                        if (this.board[i][j][2] !== this.board[start_row][start_col][2] && this.board[i][j][1].toLowerCase() !== 'k') {
                            break;
                        }
                        if (this.board[i][j][2] !== this.board[start_row][start_col][2] && this.board[i][j][1].toLowerCase() === 'k') {
                            return true;
                        }
                    }
                }
            }
        }

        if (this.piece === 'N' || this.piece === 'n') { // Knight
            for (let i = 0; i < 8; i++) {
                for (let j = 0; j < 8; j++) {
                    const delta_row = Math.abs(i - start_row);
                    const delta_col = Math.abs(j - start_col);
                    if ((delta_row === 1 && delta_col === 2) || (delta_row === 2 && delta_col === 1)) {
                        if (this.board[i][j] === null) {
                            this.board[i][j] = ['.', '.', '.'];
                        }
                        if (this.board[i][j][2] !== this.board[start_row][start_col][2] && this.board[i][j][1].toLowerCase() === 'k') {
                            return true;
                        }
                    }
                }
            }
        }
    }
}

File: s.py
class ChessGame:
    def __init__(self):
        self.board = self.initialize_board()
        self.current_player = 'White'
        self.flags = [ ['e1', 'White', 0], ['e8', 'Black', 0] ]
        self.base =  [ ['d1', 'White', 0], ['d8', 'Black', 0] ]

    def initialize_board(self):
        # Initialize an empty 8x8 chessboard with pieces in their starting positions
        board = [[None for _ in range(8)] for _ in range(8)]
        
        # Place white pieces
        board[0] = [['♖', 'R', 'Black', 0], ['♘', 'N', 'Black', 0], ['♗', 'B', 'Black', 0], ['*', 'C', 'Black', 0], ['🏴', 'K', 'Black', 0], ['♗', 'B', 'Black', 0], ['♘', 'N', 'Black', 0], ['♖', 'R', 'Black', 0]]
        board[1] = [['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0], ['♙', 'P', 'Black', 0]]

        # Place black pieces
        board[7] = [['♜', 'r', 'White', 0], ['♞', 'n', 'White', 0], ['♝', 'b', 'White', 0], ['*', 'c', 'White', 0], ['🏳️', 'k', 'White', 0], ['♝', 'b', 'White', 0], ['♞', 'n', 'White', 0], ['♜', 'r', 'White', 0]]
        board[6] = [['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0], ['♟', 'p', 'White', 0]]

        return board

    def print_board(self):
        # Print the current state of the chessboard with positions revealed
        orange_bg = '\x1b[48;5;208m'  # Orange background
        red_bg = '\x1b[48;5;196m'     # Red background
        reset_color = '\x1b[0m'        # Reset color

        print("  +------------------------ A +")
        for row in range(8):
            row_display = f"{8 - row} |"
            for col in range(8):
                if self.board[row][col] is not None:
                    piece_symbol = self.board[row][col][0]
                else:
                    piece_symbol = ' .'
                
                # Alternate background colors
                if (row + col) % 2 == 0:
                    if self.board[row][col] is None:
                        row_display += f"{orange_bg}{piece_symbol} {reset_color}"
                        self.board[row][col] = ['.','.','.']
                    elif self.board[row][col][1].lower() == 'k':
                        row_display += f"{orange_bg}{piece_symbol} {reset_color}"
                    elif self.board[row][col][1].lower() != 'k':
                        row_display += f"{orange_bg} {piece_symbol} {reset_color}"
                else:
                    if self.board[row][col] is None:
                        row_display += f"{red_bg}{piece_symbol} {reset_color}"
                        self.board[row][col] = ['.','.','.']
                    elif self.board[row][col][1].lower() == 'k':
                        row_display += f"{red_bg}{piece_symbol} {reset_color}"
                    elif self.board[row][col][1].lower() != 'k':
                        row_display += f"{red_bg} {piece_symbol} {reset_color}"
                    
            row_display += "|"
            print(row_display)
        print("  +------------------------ a +")


    # def print_board(self):
    #     # Print the current state of the chessboard with positions revealed
    #     print("  +------------------------ A +")
    #     for row in range(8):
    #         row_display = f"{8 - row} | "
    #         for col in range(8):
    #             if self.board[row][col] is not None:
    #                 piece_symbol = self.board[row][col][0]
    #             else:
    #                 piece_symbol = '.'
    #             row_display += f" {piece_symbol} "
    #         row_display += " |"
    #         print(row_display)
    #     print("  +------------------------ a +")

    def capture_piece(self, row, col, moving_piece):
        # Remove the piece at the specified position from the board if it belongs to the opposing player
        if self.board[row][col] is not None:
            self.board[row][col] = None

    def move_piece(self, piece, start_row, start_col, end_row, end_col):
        # Move the piece from start position to end position
        if self.board[start_row][start_col][1] == piece:
            self.board[end_row][end_col] = self.board[start_row][start_col]
            self.board[start_row][start_col] = ['.', '.', '.']

    def convert_position(self, position):
        # Convert algebraic notation to array indices
        row = 8 - int(position[1])
        col = ord(position[0]) - ord('a')
        return row, col

    def is_valid_move(self, piece, start_row, start_col, end_row, end_col):
        # Check if the move is within the bounds of the board
        if not (0 <= start_row < 8 and 0 <= start_col < 8 and 0 <= end_row < 8 and 0 <= end_col < 8):
            return False
        
        # Check if there is a piece at the starting position
        if self.board[start_row][start_col] is None:
            return False

        # opponent_color = 'Black' if self.board[start_row][start_col][2] == 'White' else 'White'
        # if self.is_check(opponent_color):
        #     return False  # Move leads to opponent's king being in check
        # Check for obstruction in the path
        if piece in ['R', 'r', 'Q', 'q']:  # Rook or Queen
            if start_row == end_row:  # Horizontal move
                delta_col = 1 if end_col > start_col else -1
                for col in range(start_col + delta_col, end_col, delta_col):
                    if self.board[start_row][col] is None:
                        self.board[start_row][col] = ['.', '.', '.']
                    elif self.board[start_row][col][2] == self.board[start_row][start_col][2]:
                        return False  # Obstruction found
                    elif end_col != col and self.board[end_row][col][2] != self.board[start_row][start_col][2]:
                        return False  # Obstruction found
                    elif end_col == col and self.board[end_row][col][2] != self.board[start_row][start_col][2]:
                        return True  # Obstruction found
                return True
            elif start_col == end_col:  # Vertical move
                delta_row = 1 if end_row > start_row else -1
                for row in range(start_row + delta_row, end_row, delta_row):
                    if self.board[row][start_col] is None:
                        self.board[row][start_col] = ['.', '.', '.']
                    elif self.board[row][start_col][2] == self.board[start_row][start_col][2]:
                        return False  # Obstruction found
                    elif end_row != row and self.board[row][start_col][2] != self.board[start_row][start_col][2]:
                        return False  # Obstruction found
                    elif end_row == row and self.board[row][start_col][2] != self.board[start_row][start_col][2]:
                        return True  # Obstruction found
                return True

        if piece in ['B', 'b', 'Q', 'q']:  # Bishop or Queen
            delta_row = end_row - start_row
            delta_col = end_col - start_col
            if abs(delta_row) == abs(delta_col):  # Diagonal move
                row_step = 1 if start_row < end_row else -1
                col_step = 1 if start_col < end_col else -1
                row = (start_row + row_step) % 8
                col = (start_col + col_step) % 8
                check_row = row
                check_col = col
                while row != 7 and col != 7:
                    if self.board[row][col] is None:
                        self.board[row][col] = ['.', '.', '.']
                    elif self.board[row][col][1].lower() == 'k':
                        print("Check!")
                        return True
                    elif self.board[row][col][2] == self.board[start_row][start_col][2]:
                        row_step = 1 if start_row > end_row else -1
                        col_step = 1 if start_col > end_col else -1
                    row = abs(row + row_step) % 8
                    col = abs(col + col_step) % 8

                row_step = 1 if start_row < end_row else -1
                col_step = 1 if start_col < end_col else -1
                row = (start_row + row_step) % 8
                col = (start_col + col_step) % 8
                check_row = row
                check_col = col
                while row != end_row and col != end_col:
                    if self.board[row][col] is None:
                        self.board[row][col] = ['.', '.', '.']
                    elif self.board[row][col][2] == self.board[start_row][start_col][2]:
                        row_step = 1 if start_row > end_row else -1
                        col_step = 1 if start_col > end_col else -1
                    elif self.board[row][col][2] == self.board[start_row][start_col][2]:
                        return False  # Obstruction found
                    elif row == end_row and self.board[row][col][2] == self.board[start_row][start_col][2]:
                        return False  # Obstruction found
                    elif row != end_row and self.board[row][col][2] == self.board[start_row][start_col][2]:
                        return False  # Obstruction found
                    row = abs(row + row_step) % 8
                    col = abs(col + col_step) % 8
        # Check if there's a piece of the same color at the ending position
        if self.board[end_row][end_col] is not None and self.board[end_row][end_col][2] == self.board[start_row][start_col][2]:
            return False  # Obstruction found
        
        if piece == 'N' or piece == 'n':  # Knight
            # Check if the move is a valid knight move
            delta_row = abs(end_row - start_row)
            delta_col = abs(end_col - start_col)
            # return (delta_row == 1 and delta_col == 2) or (delta_row == 2 and delta_col == 1)
            # Check if the move is a valid knight's move or teleportation
            if delta_row == 2 and delta_col == 1:
                return True
            elif delta_row == 1 and delta_col == 2:
                return True
            elif delta_row == 1 and delta_col >= 6:
                return True
            elif delta_row >= 6 and delta_col == 1:
                return True
            else:
                return False
        
        # Allow non-pawns and the king to move around the outside of the board
        if piece in ['P', 'p', 'K', 'k']:
            # Check if the move is along the edge of the board and the destination is on the opposite side
            if (start_row == 0 and end_row == 7 and start_col == end_col) or \
            (start_row == 7 and end_row == 0 and start_col == end_col) or \
            (start_col == 0 and end_col == 7 and start_row == end_row) or \
            (start_col == 7 and end_col == 0 and start_row == end_row):
                return False
            
        # Check for piece-specific move validation
        if piece == 'P':  # Pawn
            # Pawn can move forward two squares from starting position
            if start_row == 1 and start_col == end_col and end_row - start_row == 2:
                return True
            if start_col == end_col and end_row - start_row == 1:
                return True
            # Pawn can capture diagonally
            if self.board[end_row][end_col] is not None and self.board[start_row][start_col][2] != self.board[end_row][end_col][2] and abs(start_col - end_col) == 1 and end_row - start_row == 1:
                return True
            return False
        elif piece == 'p':  # Pawn (for black)
            # Pawn can move forward two squares from starting position
            if start_row == 6 and start_col == end_col and start_row - end_row == 2:
                return True
            # Similar logic for black pawn
            if start_col == end_col and start_row - end_row == 1:
                return True
            if self.board[end_row][end_col] is not None and abs(start_col - end_col) == 1 and start_row - end_row == 1:
                return True
            return False
        elif piece == 'R' or piece == 'r':  # Rook
            # Rook moves horizontally or vertically
            return start_row == end_row or start_col == end_col
        elif piece == 'B' or piece == 'b':  # Bishop
            # Bishop moves diagonally
            return abs(end_row - start_row) == abs(end_col - start_col)
        elif piece == 'Q' or piece == 'q':  # Queen
            # Queen combines rook and bishop moves
            return (start_row == end_row or start_col == end_col) or (abs(end_row - start_row) == abs(end_col - start_col))
        elif piece == 'K' or piece == 'k':  # King
            # King moves one square in any direction
            str_temp = chr(end_col + ord('a')) + end_row
            if self.board[end_row][end_col][2] == 'White':
                self.flags[0][0] = str_temp
            elif self.board[end_row][end_col][2] == 'Black':
                self.flags[1][0] = str_temp
            return abs(end_row - start_row) <= 1 and abs(end_col - start_col) <= 1
        else:
            return False  # Default: invalid move
        # Check if there's a piece of the same color at the ending position
        if self.board[end_row][end_col] is not None and self.board[end_row][end_col][2] == self.board[start_row][start_col][2]:
            return False

        return True  # Move is valid

    def is_check(self, piece, start_row, start_col, flags_row, flags_col):
        # Check if the move is within the bounds of the board
        if not (0 <= start_row < 8 and 0 <= start_col < 8 and 0 <= end_row < 8 and 0 <= end_col < 8):
            return False
        
        # Check if there is a piece at the starting position
        if self.board[start_row][start_col] is None:
            return False

        # opponent_color = 'Black' if self.board[start_row][start_col][2] == 'White' else 'White'
        # if self.is_check(opponent_color):
        #     return False  # Move leads to opponent's king being in check
        # Check for obstruction in the path
        if piece in ['R', 'r', 'Q', 'q']:  # Rook or Queen
            if start_row == flags_row:  # Horizontal move
                for i in range(8):
                    if self.board[start_row][i] is None:
                        self.board[start_row][i] = ['.', '.', '.']
                    if self.board[start_row][i][2] != self.board[start_row][i][2] and self.board[start_row][i][1].lower() != 'k':
                        return False
                    if self.board[start_row][i][2] != self.board[start_row][i][2] and self.board[start_row][i][1].lower() == 'k':
                        return True
            elif start_col == flags_col:  # Vertical move
                for i in range(8):
                    if self.board[i][start_col] is None:
                        self.board[i][start_col] = ['.', '.', '.']
                    if self.board[i][start_col][2] != self.board[start_row][start_col][2] and self.board[i][start_col][1].lower() != 'k':
                        return False
                    if self.board[i][start_col][2] != self.board[start_row][start_col][2] and self.board[i][start_col][1].lower() == 'k':
                        return True

        if piece in ['B', 'b', 'Q', 'q']:  # Bishop or Queen
            for i in range(8):
                for j in range(8):
                    if j == i:
                        if self.board[i][j] is None:
                            self.board[i][j] = ['.', '.', '.']
                        if self.board[i][j][2] == self.board[start_row][start_col][2]:
                            break
                        if self.board[i][j][2] != self.board[start_row][start_col][2] and self.board[i][j][1].lower() != 'k':
                            break
                        if self.board[i][j][2] != self.board[start_row][start_col][2] and self.board[i][j][1].lower() == 'k':
                            return True

        if piece == 'N' or piece == 'n':  # Knight
            # Check if the move is a valid knight move
            for i in range(8):
                for j in range(8):
                    delta_row = abs(i - start_row)
                    delta_col = abs(j - start_col)
                    if (delta_row == 1 and delta_col == 2) or (delta_row == 2 and delta_col == 1):
                        if self.board[i][j] is None:
                            self.board[i][j] = ['.','.','.']
                        if self.board[i][j][2] != self.board[start_row][start_col][2] and self.board[i][j][1].lower() == 'k':
                            return True

if __name__ == "__main__":
    game = ChessGame()
    current_player = 'White'  # Start with white player
    while True:
        game.print_board()
        print(f"It's {'White' if current_player == 'White' else 'Black'}'s turn.")

        move = input("Enter your move (e.g., 'N c2 to d4'): ")
        if move.lower() == 'exit':
            break

        move_parts = move.split()
        piece = move_parts[0]
        start_position = move_parts[1]
        end_position = move_parts[3]
        start_row, start_col = game.convert_position(start_position)
        end_row, end_col = game.convert_position(end_position)

        # Check if it's the player's piece
        if game.board[start_row][start_col] is None or game.board[start_row][start_col][2] != current_player:
            print("Invalid Move! It's not your piece.")
            continue

        # if game.check_endgame_modes(current_player):
        if game.is_valid_move(piece, start_row, start_col, end_row, end_col):
            # game.check_endgame_modes(current_player)
            if game.is_check(piece, start_row, start_col, end_row, end_col):
                print("Check!")
            game.move_piece(piece, start_row, start_col, end_row, end_col)

            offset = 0
            if current_player != 'White':
                offset = 1
            king_row, king_col = game.convert_position(game.flags[offset][0])

            if game.is_check(piece, end_row, end_col, king_row, king_col):
            # if game.is_check(piece, end_row, end_col, king_row, king_col):
                print("Check!")
            
            # Switch player
            current_player = 'Black' if current_player == 'White' else 'White'
        else:
            print("Invalid Move!")
